### ISSUE TITLE
Handling usage-based resources

### ISSUE THREAD

**Issue by aliscott** [infracost/infracost#302]
We currently have a [terraform provider](https://registry.terraform.io/providers/infracost/infracost/latest/docs) for adding usage estimates for usage-based resources such as AWS Lambda or S3. We’ve been thinking about how we can make it easier for the user to see usage-based cost estimates.

Some previous discussions:
* https://github.com/infracost/infracost/issues/34
* https://github.com/infracost/infracost/issues/35 
* https://github.com/infracost/infracost/issues/220

I’ve listed some preliminary ideas below and would love to get some feedback/suggestions and additional ideas for handling these.

### Idea 1: Add some default usage values 

For each usage-based resource, add some default quantities to show instead of ‘-’. For example:

```sh
 NAME                             MONTHLY QTY  UNIT         PRICE   HOURLY COST  MONTHLY COST
 aws_lambda_function.hello_world
 ├─ Requests                                1  1M requests  0.2000       0.0003        0.2000
 └─ Duration                                1  GB-seconds    2e-05        2e-08         2e-05
 Total                                                                   0.0003        0.2000
```

We can then combine this with the usage YAML file from idea 2 to allow the user to edit these.

### Idea 2: Prompt for usage data and store in a config file

If the user is using resources that need usage data then we can prompt for them the first time the user runs infracost (similar to how Terraform prompts for vars). 

For example:

```
You have 2 resources that need usage data to be estimated.
Would you like to add usage values for them now? (y/N)

Use the following values if you skip any resources
s: skip, sr: skip resource, sa: skip all

Resource: aws_s3_bucket.bucket (Standard storage class)
Storage (GB):      

Resource: aws_s3_bucket.bucket (Standard storage class)
PUT, COPY, POST, LIST requests (requests):     

Your usage data has been saved in infracost-usage.yml. You can update this and rerun `infracost`.
```

We would then store these usage values in a file in the current directory so they can be updated by the user and used for each subsequent run.

infracost-usage.yml
```yaml
aws_s3_bucket.bucket2:
  Standard:
    "Storage (GB)": 250
    "PUT, COPY, POST, LIST requests (requests)": 100000
```

###  Idea 3: Pull usage data for the last month from the cloud vendor

This one needs some more investigation, but the idea would be to query the cloud vendor APIs to retrieve usage data for each resource for the last month (where possible) and use that to populate the Monthly Quantity column. Using resource IDs or tags from Terraform might help to find usage data.

**Comment by jgrumboe** [comment]
Hi Ali,

I think default values for usage are always a little bit tricky, so I wouldn't do that.

Option 3 would be cool, but it wouldn't probably reflect cost-changes which would be generated by new stuff configured. Let's assume an update to an existing lambda function which needs more memory - last month usage wouldn't reflect that, so infracost would probably tell me no cost increase will happen.

I think option 2 would be the best way to get. I like the idea that it would prompt in terminal mode for usage parameters and saving this information into a configfile, which can be reused for CI.
Probably an "update my usage from cloud vendor" option and updating the config-file would be a great addition.

Greetings,
Jo

**Comment by shomeprasanjit** [comment]
Ali, 
    I think a combination of option 1 and option 2 is ideal, the following are the benefits :
    
    1) Users don't get to see '-' if no usage data is passed from the user side. 
    2) Allow new users to be up and running with this tool without much tweaking and at the same time, they understand these values can be tweaked anytime using the config file.

**Comment by aliscott** [comment]
Thanks @jgrumboe and @shomeprasanjit I'm investigating doing #2 as part of https://github.com/infracost/infracost/issues/220 so will update here when I have a version to play with.

**Comment by alikhajeh1** [comment]
Option 2 has been released in [v0.7.12](https://github.com/infracost/infracost/releases/tag/v0.7.12) (also available on homebrew) so I'm going to close this issue, @jgrumboe and @shomeprasanjit feel free to leave any feedback here or on the community Slack channel. New issues can also be created if there are any bugs with it or there are ideas on how to improve it (e.g. does it work ok for bigger projects).

[This file](https://github.com/infracost/infracost/blob/master/infracost-usage-example.yml) shows the currently supported resources, so you can copy/paste the sections you need from that and update the resource names/values, then run `infracost --usage-file /path/to/infracost-usage.yml`.
